### PR TITLE
Add targets

### PR DESCRIPTION
**Pull request by SekoiaTree**
This PR adds support for targets, which allow a document to know what it's being compiled into. 

This is part of the basics that will eventually allow HTML compilation.

Currently missing:
- tests (might not be needed? They don't really work with the existing testing pipeline, since they're meta)
- clippy. So much clippy (probably)

### CODE REVIEW

File: crates/typst-library/src/meta/target.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#[func]
+pub fn target(
+    /// The virtual machine.
+    vm: &mut Vm,
+) -> Target {
+    Target::new(vm.vt.world.target())
+}
+
+#[derive(Clone, PartialEq, Hash, Debug)]
+pub struct Target(export::Target);
```
Comment (PgBiel) [inline crates/typst-library/src/meta/target.rs:52]: The `Debug` (and, thus, `repr()`) impl should probably just provide a string version of `Target` (e.g. `"pdf"`, `"png"`, `"svg"`, `"query"`).

Comment (istudyatuni) [reply crates/typst-library/src/meta/target.rs:52]: As I understand Debug, it's about internal representation

Comment (PgBiel) [reply crates/typst-library/src/meta/target.rs:52]: > As I understand Debug, it's about internal representation

It's used by `repr` for user-facing value types, and thus doesn't normally expose the internal Rust structure (as is, you'd see something like `Target(Png)` or similar, which isn't consistent with other types).
https://github.com/typst/typst/blob/88bcb8769a07d1bad57b77fa9443b598fd1a6b61/crates/typst/src/eval/value.rs#L143-L146

**Comment by reknih** [comment]
In my opinion, it should not be PNG but `raster` or similar. We create a pixel buffer, the file format / compression is incidental. For example, we could also produce a TIFF or JPEG but the target considerations should be the same.

**Comment by SekoiaTree** [comment]
Should I change SVG to `vector` or similar, then?

File: crates/typst-library/src/meta/target.rs (hunk 1 above)
Comment (SekoiaTree) [reply crates/typst-library/src/meta/target.rs]: Fixed.

**Comment by SekoiaTree** [comment]
I ended up changing `Png` to `Raster` and `Svg` to `Vector`. Pdf stays the same since it's really the only format that renders... like that.

**Comment by LaurenzV** [comment]
To be honest, I'm personally somewhat against having such a feature in Typst, at least at the current stage. Because by introducing the possibility to add target-specific code, we can no longer expect all of the output formats to look the same, which in my opinion is an unfortunate side-effect, especially for the web app, since then the outputted PDF might look differently than what was shown in the preview.

I'm aware that since HTML support is planned, at some point there might have to be a way to introduce target-specific code (although I have no clue how this could be done in a nice way), but at least for now ,while we still have the same frame-based representation for each output format (PNG, PDF, SVG), I would personally steer away from allowing users to specify target-specific code.

Just my two cents though. 🙂

**Comment by PgBiel** [comment]
We could perhaps delay this change to when HTML export is added, and maybe have a way to differentiate HTML export from the rest. Regarding differentiating PDF / PNG / SVG, that could be postponed until we have a proper discussion on whether that's worth it or not.

**Comment by laurmaedje** [comment]
I originally thought that there was no harm in supporting this as is already, but it is true that there's still a discussion to be had how fine-grained targets should be. It could be per-format with simple accessors for groups of related formats or it could be that we only expose the group information. We could even store the format internally, but not expose it and that way still have overlapping groups of formats (like is-layouted, is-x, is-y).

**Comment by laurmaedje** [comment]
I'll close this for now as things will still take a bit. We can see again then. I might also just add this alongside other stuff when I'm working on the relevant parts.

**Pull request closed without merging**